File: whatsapp_bot/build_vector_db.py
#!/usr/bin/env python3
import json
import os
import sys
import numpy as np
from sentence_transformers import SentenceTransformer

# Paths
PWD = os.path.dirname(os.path.abspath(__file__))
INPUT_PATH = os.path.join(PWD, 'rag_database_refined.json')
OUTPUT_PATH = os.path.join(PWD, 'rag_vectors.json')

def main():
    print("[RAG Build] Loading multilingual model...")
    model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')

    print(f"[RAG Build] Loading refined database from {INPUT_PATH}...")
    if not os.path.exists(INPUT_PATH):
        print(f"[ERROR] {INPUT_PATH} not found. Run refine_rag_data.js first.")
        return

    with open(INPUT_PATH, 'r', encoding='utf-8') as f:
        data = json.load(f)

    print(f"[RAG Build] Loaded {len(data)} chunks. Starting vectorization...")
    
    # Process in batches for speed
    BATCH_SIZE = 128
    vectors_output = []

    for i in range(0, len(data), BATCH_SIZE):
        batch = data[i:i + BATCH_SIZE]
        try:
            # Generate embeddings
            embeddings = model.encode(batch, convert_to_numpy=True, normalize_embeddings=True)
            
            for text, vec in zip(batch, embeddings):
                vectors_output.append({
                    "text": text,
                    "vector": vec.tolist()
                })
            
            if i % 1280 == 0:
                print(f"  Processed: {i} / {len(data)}")
        except Exception as e:
            print(f"  [ERROR] Processing batch starting at {i}: {e}")

    print(f"[RAG Build] Saving {len(vectors_output)} vectors to {OUTPUT_PATH}...")
    with open(OUTPUT_PATH, 'w', encoding='utf-8') as f:
        json.dump(vectors_output, f)

    print("[SUCCESS] RAG memory is fully rebuilt and optimized.")

if __name__ == "__main__":
    main()

File: whatsapp_bot/build_vector_db.js
const fs = require('fs');
const path = require('path');

async function main() {
    console.log("Loading Multilingual Transformer Pipeline natively in Node...");
    // Dynamic import to support ESM modules
    const { pipeline } = await import('@xenova/transformers');
    
    console.log("Loading Rag Database...");
    let ragDatabase = [];
    try {
        ragDatabase = JSON.parse(fs.readFileSync(path.join(__dirname, 'rag_database_refined.json'), 'utf8'));
    } catch(e) {
        console.error("No database found!");
        return;
    }
    
    console.log(`Loaded ${ragDatabase.length} records. Prepping native Multilingual AI pipeline...`);
    // Load multilingual model
    // This explicitly protects "Hinglish" and Indian cultural semantics
    const extractor = await pipeline('feature-extraction', 'Xenova/paraphrase-multilingual-MiniLM-L12-v2', {
        quantized: true // Use INT8 to speed up the embedding exponentially
    });
    
    console.log("Model loaded. Vectorizing 60k strings in batches...");
    const vectors = [];
    const BATCH_SIZE = 64;
    
    for (let i = 0; i < ragDatabase.length; i += BATCH_SIZE) {
        const batch = ragDatabase.slice(i, i + BATCH_SIZE);
        try {
            // Extractor can process a batch of strings
            const outputs = await extractor(batch, { pooling: 'mean', normalize: true });
            
            // Map outputs back to our vector structure
            for (let j = 0; j < batch.length; j++) {
                const vectorArray = Array.from(outputs[j].data);
                vectors.push({
                    text: batch[j],
                    vector: vectorArray
                });
            }
            
            if (i % 640 === 0) {
                console.log(`Processed: ${i} / ${ragDatabase.length}`);
            }
        } catch(e) {
            console.error(`Error processing batch starting at ${i}:`, e);
        }
    }
    
    console.log(`Vectorization complete. Saving ${vectors.length} semantic vectors to rag_vectors.json...`);
    fs.writeFileSync(path.join(__dirname, 'rag_vectors.json'), JSON.stringify(vectors));
    console.log("Saved.");
}

main();
